text_extensions_for_pandas/array/span.py

```diff
@@ -382,43 +457,60 @@ def __getitem__(self, item) -> Union[Span, "SpanArray"]:
         for information about this method.
         """
         if isinstance(item, int):
-            return Span(self._text, int(self._begins[item]),
+            return Span(self.target_text[item], int(self._begins[item]),
                         int(self._ends[item]))
         else:
             # item not an int --> assume it's a numpy-compatible index
             item = check_array_indexer(self, item)
-            return SpanArray(self._text,
-                             self._begins[item],
-                             self._ends[item])
+            return SpanArray(
+                (self._string_table, self._text_ids[item]),
+                self._begins[item], self._ends[item])
 
     def __setitem__(self, key: Union[int, np.ndarray], value: Any) -> None:
         """
         See docstring in `ExtensionArray` class in `pandas/core/arrays/base.py`
         for information about this method.
         """
+        # Subroutine of the if-else sequence below
+        def _is_sequence_of_spans(seq: Any):
+            if isinstance(seq, SpanArray):
+                return True
+            if not isinstance(seq, (collections.abc.Sequence, np.ndarray)):
+                return False
+            else:
+                # For other sequences, check for everything being Span or None
+                for elem in seq:
+                    if elem is not None and not isinstance(elem, Span):
+                        return False
```

Comment:
Just a suggestion, what is here is fine too

```suggestion
                if any(elem is not None and not isinstance(elem, Span) for elem in seq):
                    return False
```

Reply:
Good idea. Shortened further to
```python
                return all(elem is not None and not isinstance(elem, Span)
                           for elem in seq)
```